### PR TITLE
Bugfix/speed

### PR DESCRIPTION
**Pull request by heeler**
The goal of this branch was to change how memory for the numpy array was handled more efficiently. 

The memory container is the ImagesContainer class that returns the typed memory via polymorphism. 
Pointers from the single block of memory are used to construct TypedImages with references to where their image should be written. The TypedImage then copies the bitmap to the memory TypedImage is pointing to.

The code changes rely on polymorphism to handle the different types of memory. This is a lot more complicated than casting a void * and maybe a little nutty but I find it a bit scary to have exposed code that requires the memory to be casted to the right type. This approach means all the logic is self-contained in the relevant classes. I'd love your thoughts on this design style.

If you have suggestions on commenting, coding style, design or any other points I'd really appreciate it.

### CODE REVIEW

File: _aicspylibczi/ImageFactory.cpp
Comment (toloudis) [inline _aicspylibczi/ImageFactory.cpp:67]: delete this block?

File: _aicspylibczi/ImageFactory.cpp
Comment (toloudis) [inline _aicspylibczi/ImageFactory.cpp:133]: Something feels weird about how s_pixelToImage is named.   
Maybe renaming it to something like s_pixelTypeToImageConstructor, and/or :
```
auto imageFactoryFunction = s_pixelToImage[pixelType];
std::shared_ptr<Image> image = imageFactoryFunction(...)
```

Comment (toloudis) [reply _aicspylibczi/ImageFactory.cpp:133]: or even something like s_getImageFactoryFunction, just more descriptive of its usage.

File: _aicspylibczi/ImageFactory.cpp
Comment (toloudis) [inline _aicspylibczi/ImageFactory.cpp:138]: Just wondering, is this line needed?  shared_ptr should increment its ref on copy.

File: _aicspylibczi/ImageFactory.h
Comment (toloudis) [inline _aicspylibczi/ImageFactory.h:24]: It looks like this class was formerly 100% static.  Why does it need this member variable now?  Could the imgContainer be passed in from the outside to keep all the member functions static?

File: _aicspylibczi/ImagesContainer.h
Comment (toloudis) [inline _aicspylibczi/ImagesContainer.h:32]: this should probably be dynamic_cast since it is actually a derived class... in fact why not just call dynamic_cast instead of getBaseAsTyped?

File: _aicspylibczi/ImageFactory.cpp
Comment (JacksonMaxfield) [inline _aicspylibczi/ImageFactory.cpp:159]: I would love more comments in general but specifically around here because next time we have bugs with BGR images this will be the first place to look I would assume.

File: _aicspylibczi/Reader.cpp
Comment (JacksonMaxfield) [inline _aicspylibczi/Reader.cpp:19]: It's impossible to have varied pixel types by scene right? I feel like I have personally confirmed this before but just double checking.

Comment (JacksonMaxfield) [reply _aicspylibczi/Reader.cpp:19]: I'm like 99% sure we have checked this assumption before.... just can't find it in slack.

File: _aicspylibczi/Reader.cpp
Comment (JacksonMaxfield) [inline _aicspylibczi/Reader.cpp:223]: Python convention would be to place this function above when it is used don't know about C though.

File: aicspylibczi/tests/test_czi_file.py
Comment (JacksonMaxfield) [inline aicspylibczi/tests/test_czi_file.py:252]: Why remove the `1` here?

File: c_tests/test_Reader.cpp
Comment (JacksonMaxfield) [inline c_tests/test_Reader.cpp:56]: These should probably point to actual test files?

File: _aicspylibczi/Reader.cpp
Comment (toloudis) [reply _aicspylibczi/Reader.cpp:223]: class member functions can go anywhere

File: _aicspylibczi/Reader.cpp
Comment (toloudis) [inline _aicspylibczi/Reader.cpp:273]: there is something strange about this calling constructImage and returning an image and then never using it.  It has to be a naming problem, I suppose.  What is constructImage really doing?

File: _aicspylibczi/Reader.h
Comment (toloudis) [inline _aicspylibczi/Reader.h:284]: could this be initialized earlier in order to keep the method const?

File: _aicspylibczi/pb_caster_ImagesContainer.h
Comment (toloudis) [inline _aicspylibczi/pb_caster_ImagesContainer.h:14]: comment needs updating

File: main.cpp
Comment (toloudis) [inline main.cpp:12]: another overly specific path

**Comment by toloudis** [comment]
Some general thoughts I will offer.  Some of this is opinion and some of it universal: 

1.  There's not enough whitespace in some areas.  Whitespace is good for readability.  I don't know if you are using a formatter, but that can help a lot.  I use one that auto-formats on save in my editor.   
1a. For example, I got into the habit of putting the consequent of every "if" block into curly brackets {} and on a new line.  It also makes it safer if I need to add a second line to the consequent.
1b. Second example, I prefer to see spaces around operators that take two variables (like &&, ==, =, etc)

2. I think some of the variable names should be a little more verbose and descriptive.  For some of this code I think better variable naming would go a very long way for readability.

3. This code is complicated for someone trying to read it coming in cold.   Comments about the intended behavior/actions of several of these functions would really help too.  Even if I don't do a full docstring type of comment, I like to put a comment above the function just saying what its intention is -- especially if the name is hard to make descriptive enough.

File: _aicspylibczi/pb_caster_ImagesContainer.h
Comment (JacksonMaxfield) [reply _aicspylibczi/pb_caster_ImagesContainer.h:14]: yes more comments

File: _aicspylibczi/ImageFactory.cpp
Comment (heeler) [reply _aicspylibczi/ImageFactory.cpp:138]: I would have used a unique_ptr but you can't use the polymorphic trick with unique_ptrs whereas you can with shared pointers. The other thing is that I wanted to add it to the m_imgContainer member and return it to the caller in case they wanted to structure the images differently. 

Does that address your concern or am I missing something?

File: _aicspylibczi/ImageFactory.h
Comment (heeler) [reply _aicspylibczi/ImageFactory.h:24]: My thinking was to let the ImageFactory handle all the logic of memory, essentially making it a controller and to keep any memory logic (outside of the pybind11 object conversion) contained in the ImageFactory. Is there an advantage to having all the functions being static? I assumed that essentially just mean the class is a namespace?

File: _aicspylibczi/ImagesContainer.h
Comment (heeler) [reply _aicspylibczi/ImagesContainer.h:32]: Should I be using dynamic_cast or dynamic_pointer_cast? I'm reading the docs and trying to figure it out but it looks like there's a subtlety I think I'm missing.

File: _aicspylibczi/Reader.cpp
Comment (heeler) [reply _aicspylibczi/Reader.cpp:19]: You're going to hate this I know I do. From digging through all the docs I have all I could find is that PixelType has to be consistent across bitmaps which to me says in a "normal" image subblocks could have different PixelTypes but in Mosiac files, they have to be the same for the different M-indexes for the same dims. 

In practice, I can't think how or why one would change PixelType for different parts of an image acquisition. 
I think this is just base file format flexibility that is present but not used or at least not used on pyramid0.

File: _aicspylibczi/ImageFactory.cpp
Comment (toloudis) [reply _aicspylibczi/ImageFactory.cpp:138]: The variable `image` is already a shared_ptr.  So you can just pass it around by value without creating the `sImage` variable.  That's what I was getting at.

File: _aicspylibczi/ImageFactory.h
Comment (toloudis) [reply _aicspylibczi/ImageFactory.h:24]: Yes, it has huge implications for whether you have to instantiate the class or can just call its functions without creating objects.  But it all depends on your desired usage pattern.

File: _aicspylibczi/Reader.cpp
Comment (heeler) [reply _aicspylibczi/Reader.cpp:273]: It's actually internally adding it to a vector but returns it to the user in case they want to do something with it. I don't have to assign the return value? What's your preferred behavior here?

File: _aicspylibczi/ImagesContainer.h
Comment (toloudis) [reply _aicspylibczi/ImagesContainer.h:32]: It looked to me like `getBaseAsTyped` could be completely replaced with `dynamic_cast`.  Or if you feel more comfortable, replace its internal static_cast with dynamic_cast.   `dynamic_pointer_cast` is relevant for `shared_ptr`s only and is not pertinent for this code.

File: _aicspylibczi/Reader.h
Comment (heeler) [reply _aicspylibczi/Reader.h:284]: I'm essentially lazy loading here, I pushed things like this off so that we weren't taking performance hits for all the "we should initialize these up front". I have to have a think on this.

File: main.cpp
Comment (heeler) [reply main.cpp:12]: So main.cpp was written solely for profiling against larger files. Should I just remove it from the project or what do you think the right thing to do is?

**Comment by heeler** [comment]
Thanks for the general comments on white space and coding style @toloudis. It's probably a bad habit but I've somewhat evolved a "compact"/dense coding style but you're right it makes it hard for others to read my stuff. Thanks for pointing it out. I should probably just configure the clang formatter.

File: _aicspylibczi/ImageFactory.cpp
Comment (heeler) [reply _aicspylibczi/ImageFactory.cpp:138]: Thanks I totally didn't see that but yes your 100% right. thanks for catching that.

File: _aicspylibczi/ImagesContainer.h
Comment (heeler) [reply _aicspylibczi/ImagesContainer.h]: So there's a template/polymorphism clash happening here so I think I'm stuck with static_cast.

File: main.cpp
Comment (toloudis) [reply main.cpp:12]: It's a fair point.  One alternative is passing that file path as a command line arg.

File: _aicspylibczi/ImagesContainer.h
Comment (heeler) [reply _aicspylibczi/ImagesContainer.h]: The trick to getting this going was as @toloudis suggested, adding a virtual destructor to the base class. Then polymorphism with the templated child class works with dynamic_cast.

File: _aicspylibczi/ImageFactory.cpp
Comment (toloudis) [inline _aicspylibczi/ImageFactory.cpp:83]: delete this?

File: _aicspylibczi/ImageFactory.h
Comment (toloudis) [inline _aicspylibczi/ImageFactory.h:31]: is it clear enough to a caller that this function has side effects on the object?  Should it be called something like relinquish or transferMemory?

File: _aicspylibczi/Reader.cpp
Comment (toloudis) [reply _aicspylibczi/Reader.cpp:273]: I guess the preferred thing is to drop the `auto image = ` part of the line.

File: _aicspylibczi/Reader.cpp
Comment (toloudis) [inline _aicspylibczi/Reader.cpp:328]: delete the above commented out lines, as it looks like they are meant to be replaced by the code below it

File: _aicspylibczi/Reader.cpp
Comment (toloudis) [inline _aicspylibczi/Reader.cpp:412]: again it looks like you don't need the `auto image = ` part of this line.  the `image` var is never being used.

File: _aicspylibczi/SourceRange.h
Comment (toloudis) [inline _aicspylibczi/SourceRange.h:62]: This suddenly looks odd to me.  a variable called `m_ptr` is actually a vector of pointers?

File: _aicspylibczi/TypedImage.h
Comment (toloudis) [inline _aicspylibczi/TypedImage.h:56]: when I am passing around memory offsets, I try to prefer size_t over int

File: _aicspylibczi/SourceRange.h
Comment (heeler) [reply _aicspylibczi/SourceRange.h:62]: so you want a better name? m_vPtr or m_ptrVec ?

Comment (toloudis) [reply _aicspylibczi/SourceRange.h:62]: Yes -- how about m_channelPtrs or even m_channels?

File: _aicspylibczi/ImageFactory.h
Comment (heeler) [reply _aicspylibczi/ImageFactory.h:24]: I get what you are saying. In this case, I don't think it has much impact. The alternative is to add the memory object to every function call and the cost is the class being stateful but it is only being used in a stateful context anyway.

**Review by JacksonMaxfield — APPROVED** [review]
Looks good to me. I would wait for Dan's approval though since he is the expert here.

File: _aicspylibczi/Image.h
Comment (toloudis) [inline _aicspylibczi/Image.h:72]: the virtual destructor should be at the top of the inheritance hierarchy to avoid memory leaks.  at least SubblockSortable, or whatever base class it has.

File: _aicspylibczi/Reader.cpp
Comment (toloudis) [inline _aicspylibczi/Reader.cpp:123]: it looks like this file didn't get run through the formatter?

Comment (heeler) [reply _aicspylibczi/Reader.cpp:123]: Actually it did. I just ran it through the JetBrains default but it doesn't appear to be as specific as clang-format. I'll find a clang-format file that is decent and try applying that.

Comment (toloudis) [reply _aicspylibczi/Reader.cpp:123]: are these consistent whitespace?  `if (!isMosaic() && (!sceneBool || sceneSize==1))`  vs `if (kVal.first==DimIndex::S)` ? If so, then I'm good.

File: main.cpp
Comment (toloudis) [reply main.cpp:12]: if you make this path a parameter and just stash the path to your favorite file in a comment or something, then other users will have a chance to run this

File: _aicspylibczi/Reader.cpp
Comment (heeler) [reply _aicspylibczi/Reader.cpp:123]: So I added a .clang-format config with Mozilla presets. Apparently this is Mozilla formatted. Logical operations like == != etc No spaces but && and || are always padded with spaces. So yes it's a consistent style.

File: _aicspylibczi/Image.h
Comment (heeler) [reply _aicspylibczi/Image.h]: Thanks for catching that.

**Review by toloudis — APPROVED** [review]

**Pull request merged**
